clients/drcachesim/tests/snoop_filter_test.cpp

```diff
@@ -0,0 +1,205 @@
+ * Copyright (c) 2016-2023 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// Unit tests for snoop filter.
+#include <iostream>
+#undef NDEBUG
+#include <assert.h>
+#include "snoop_filter_test.h"
+#include "simulator/cache_lru.h"
+#include "snoop_filter.h"
+#include "cache.h"
+#include <algorithm>
+#include <memory>
+
+enum {
+    CPU_0,
+    CPU_1,
+    CPU_2,
+    CPU_3,
+};
+
+static const addr_t ADDR_A = 128;
+static const int SNOOPED_FILTER_LINE_SIZE = 32;
+
+class snoop_filter_test_t {
+
+public:
+    snoop_filter_test_t(const int num_cores)
+        : num_cores_(num_cores)
+    {
+    }
+    ~snoop_filter_test_t()
+    {
+        if (snooped_caches_ != NULL) {
+            delete[] snooped_caches_;
+        }
+        if (snoop_filter_ != NULL) {
+            delete snoop_filter_;
+        }
+    }
+
+    void
+    initialize_caches_and_snoop_filter()
+    {
+        cache_t *llc = new cache_lru_t;
+        if (!llc->init(/*associatibity=*/8, /*line_size=*/64,
+                       /*total_size=*/1024, nullptr,
+                       new cache_stats_t((int)64, "", true))) {
+            std::string error_string =
+                "Usage error: failed to initialize LL cache.  Ensure sizes and "
+                "associativity are powers of 2, that the total size is a multiple "
+                "of the line size, and that any miss file path is writable.";
```

Comment:
But your own code here is setting all of these parameters: so this error message doesn't really make sense as the caller/user cannot ensure any of these mentioned things.